### PR TITLE
Match LayoutUpdated and ScrollChanged event behavior to WPF/UWP

### PR DESCRIPTION
**Pull request by grokys**
## What does the pull request do?

A few changes that are needed for compatibility with WPF/UWP that are causing problems with `ItemsRepeater`:

- Raise `Layoutable.LayoutUpdated` at the end of layout pass: this attempts to match the behavior of `LayoutUpdated` in WPF and UWP. Previously `LayoutUpdated` was raised at the end of each control's `Arrange` method. Instead, make `LayoutManager` raise a `LayoutUpdated` event at the end of a layout and trigger a control's `LayoutUpdated` event off this
- Raise `ScrollViewer.ScrollChanged` at the end of layout pass: instead of raising `ScrollChanged` each time one of the relevant properties changes, instead raise it at the end of a layout pass. Again, this matches WPF's behavior.

## Checklist

- [x] Added unit tests (if possible)?

### CODE REVIEW

File: tests/Avalonia.Layout.UnitTests/LayoutableTests.cs [hunk 1]
```diff
@@ -203,6 +203,127 @@ public void Attaching_Control_To_Tree_Invalidates_Parent_Measure()
...
+        {
+            Border border1;
+            Border border2;
+            var layoutManager = new LayoutManager();
+            var root = new TestRoot
+            {
+                Child = border1 = new Border
+                {
+                    Child = border2 = new Border
+                    {
```
Comment (MarchingCube) [inline tests/Avalonia.Layout.UnitTests/LayoutableTests.cs:217]: Leftover brackets?

File: src/Avalonia.Controls/ScrollViewer.cs [hunk 2]
```diff
@@ -602,5 +583,35 @@ protected override void OnKeyDown(KeyEventArgs e)
...
+        protected virtual void OnScrollChanged(ScrollChangedEventArgs e)
+        {
+            RaiseEvent(e);
+        }
+
+        private void OnLayoutUpdated(object sender, EventArgs e) => RaiseScrollChanged();
+
+        private void RaiseScrollChanged()
+        {
+            var e = new ScrollChangedEventArgs(
```
[the remark this answers is not in the document]
Comment (grokys) [reply src/Avalonia.Controls/ScrollViewer.cs:605]: Good point!

[the remark this answers is not in the document]
Comment (grokys) [reply src/Avalonia.Controls/ScrollViewer.cs]: Fixed.

File: src/Avalonia.Controls/ScrollViewer.cs [hunk 3]
```diff
@@ -602,5 +583,38 @@ protected override void OnKeyDown(KeyEventArgs e)
...
+
+        private void OnLayoutUpdated(object sender, EventArgs e) => RaiseScrollChanged();
+
+        private void RaiseScrollChanged()
+        {
+            var extentDelta = new Vector(Extent.Width - _oldExtent.Width, Extent.Height - _oldExtent.Height);
+            var offsetDelta = Offset - _oldOffset;
+            var viewportDelta = new Vector(Viewport.Width - _oldViewport.Width, Viewport.Height - _oldViewport.Height);
+
+            if (extentDelta != default || offsetDelta != default || viewportDelta != default)
```
Comment (MarchingCube) [inline src/Avalonia.Controls/ScrollViewer.cs:609]: I think to have correct behavior (numerically) you might want to use something like `extentDelta.IsNearlyEquals(default)` perhaps.

**Review by MarchingCube — APPROVED** [review]
Apart from that bitwise equals check problem - LGTM.

File: src/Avalonia.Controls/ScrollViewer.cs (hunk 3 above)
Comment (danwalmsley) [reply src/Avalonia.Controls/ScrollViewer.cs]: added a commit to fix this.

**Pull request merged**